README.md
# mplotutils

helper functions for cartopy and matplotlib


## Versions

### 0.3.0

 * Added ``nrow`` and ``ncol`` parameters to ``set_map_layout`` for use with a
   gridspec.
 * Formatted with black and isort, checked with flake8.


### 0.2.0 (01.06.2018)

Mayor release, mostly introducing the new `colorbar` functionality.

 * add `_color_palette` that selects colors from the whole range of the colormap. Previously we used the `seaborn` logic that excludes colors at the edge of the colorbar (closes GH:#1).
 * the colorbar functionality was entirely redesigned:
   * new top-level function `mpu.colorbar`
   * added `aspect` keyword which defines the ratio of long to short side
   * the default width/ height is now `aspect=20` (was `size=0.04`)
   * `pad` and `size` are now scaled by the width/ height of the axes (was width/ height of the figure)
   * `pad` now uses the matplotlib default (0.05 for vertical bars and 0.15 for horizontal colorbars)
 * add `_is_monotonic` to `infer_interval_breaks`
 * renamed `sample_data_3d` to `sample_data_map`

### 0.1.0 (13.03.2018)

 * Functionality as developped for [pyvis](https://github.com/C2SM/pyvis/) (see below), packed into a module.


## History

This package bases on functions developped for the [python visualisation workshop at C2SM](https://github.com/C2SM/pyvis/) (Part 3 -> `utils.py`).
`utils.py` in turn is based on [mutils](https://github.com/mathause/mutils). 


CHANGELOG.md
# Changelog

## v0.4.0 (unreleased)

### Deprecations

### Enhancements

### Bug fixes

### Internal changes


## v0.3.1 (09.02.2023)

Version v0.3.1 fixes a regression and an additional bug not caught in from v0.3.0. It's
recommended to use this version.

### Bug fixes

- Fixed a bug in `set_map_layout`: the data ratio of a cartopy `GeoAxesSubplot` requires
  a `draw` to be correct ([#61](https://github.com/mathause/mplotutils/pull/61)).
- Fix a regression introduced in [#33](https://github.com/mathause/mplotutils/pull/33):
  `cyclic_dataarray` now correctly extrapolates the coordinates
  ([#58](https://github.com/mathause/mplotutils/pull/58)).


## v0.3.0 (15.01.2023)

Version 0.3.0 includes some long-overdue bug fixes, modernization of the code, much better
test coverage and some enhancements.

mplotutils now uses the MIT license instead of GPL-3.0 ([#51](https://github.com/mathause/mplotutils/pull/51))

### Deprecations

 * Deprecated `mpu.infer_interval_breaks` as this is no longer necessary with matplotlib v3.2
   and cartopy v0.21 ([#32](https://github.com/mathause/mplotutils/pull/32)).
 * Deprecated a number of positional arguments, these are now keyword only, e.g. in
   `mpu.colorbar` ([#54](https://github.com/mathause/mplotutils/pull/54)).

### Enhancements

 * Added ``nrow`` and ``ncol`` parameters to ``set_map_layout`` for use with a
   gridspec.
 * Allow passing list of axes to ``set_map_layout``, renamed the files and extended
   the test coverage ([#42](https://github.com/mathause/mplotutils/pull/42)
   and [#43](https://github.com/mathause/mplotutils/pull/43)).
 * Add function to create `xr.DataArray` sample data ([#53](https://github.com/mathause/mplotutils/pull/53)).

### Bug fixes

 * Fixed compatibility of `mpu.colorbar` with `bbox_inches="tight"` for matplotlib 3.4 and
   newer and refactor and extend tests ([#26](https://github.com/mathause/mplotutils/pull/26)
   and [#40](https://github.com/mathause/mplotutils/pull/40)).

### Internal changes

 * Replaced `ax.get_geometry()` with `ax.get_subplotspec().get_geometry()`
   as the former was deprecated in matplotlib ([#8](https://github.com/mathause/mplotutils/pull/8)).
 * Refactor `mpu.cyclic_dataarray` using `obj.pad` ([#33](https://github.com/mathause/mplotutils/pull/33)).
 * Enabled CI on github actions ([#9](https://github.com/mathause/mplotutils/pull/9)).
 * Formatted with black and isort, checked with flake8.

## v0.2.0 (01.06.2018)

Mayor release, mostly introducing the new `colorbar` functionality.

 * add `_color_palette` that selects colors from the whole range of the colormap. Previously
   we used the `seaborn` logic that excludes colors at the edge of the colorbar
   ([#1](https://github.com/mathause/mplotutils/issues/1)).
 * the colorbar functionality was entirely redesigned:
   * new top-level function `mpu.colorbar`
   * added `aspect` keyword which defines the ratio of long to short side
   * the default width/ height is now `aspect=20` (was `size=0.04`)
   * `pad` and `size` are now scaled by the width/ height of the axes (was width/ height of the figure)
   * `pad` now uses the matplotlib default (0.05 for vertical bars and 0.15 for horizontal colorbars)
 * add `_is_monotonic` to `infer_interval_breaks`
 * renamed `sample_data_3d` to `sample_data_map`

## v0.1.0 (13.03.2018)

 * Functionality as developped for [pyvis](https://github.com/C2SM/pyvis/) (see below), packed into a module.


mplotutils/test/test_label_map.py
from mplotutils import ylabel_map, xlabel_map

import numpy as np
import matplotlib.pyplot as plt

# test it does not error...
def test_works():


	f, ax = plt.subplots()
	ylabel_map('ylabel', ax=ax)
	xlabel_map('ylabel', ax=ax)






mplotutils/test/test_colorbar_utils.py

from pytest import raises

from mplotutils.colorbar_utils import (colorbar,
                                       _resize_colorbar_vert,
                                       _resize_colorbar_horz,
                                       _parse_shift_shrink,
                                       _parse_size_aspect_pad,
                                       _get_cbax)

import matplotlib.pyplot as plt
import numpy as np


def test_parse_shift_shrink():

    # test code for _parse_shift_shrink
    assert _parse_shift_shrink('symmetric', None) == (0., 0.)

    assert _parse_shift_shrink('symmetric', 1) == (0.5, 1)
    
    assert _parse_shift_shrink(1, None) == (1, 1)
    
    assert _parse_shift_shrink(1, 1) == (1, 1)

    assert _parse_shift_shrink(0.5, 0.5) == (0.5, 0.5)

    with raises(AssertionError):
        _parse_shift_shrink(-0.1, 0)

    with raises(AssertionError):
        _parse_shift_shrink(1.1, 0)

    with raises(AssertionError):
        _parse_shift_shrink(0, -0.1)

    with raises(AssertionError):
        _parse_shift_shrink(0, 1.1)

# =============================================================================

def test_parse_size_aspect_pad():
    """
    size, aspect, pad = _parse_size_aspect_pad(size, aspect, pad, 'horizontal')
    """

    res = _parse_size_aspect_pad(0.1, None, 0.1, 'horizontal')
    exp = (0.1, None, 0.1)
    assert res == exp

    res = _parse_size_aspect_pad(None, None, 0.1, 'horizontal')
    exp = (10, 20, 0.1)
    assert res == exp

    res = _parse_size_aspect_pad(None, 20, 0.1, 'horizontal')
    exp = (10, 20, 0.1)
    assert res == exp

    with raises(ValueError):
        _parse_size_aspect_pad(1, 1, 0.1, 'horizontal')

    res = _parse_size_aspect_pad(None, None, None, 'horizontal')
    exp = (10, 20, 0.15)
    assert res == exp

    res = _parse_size_aspect_pad(None, None, None, 'vertical')
    exp = (10, 20, 0.05)
    assert res == exp



# =============================================================================


def _easy_cbar_vert(**kwargs):

    f, ax = plt.subplots()
    
    f.subplots_adjust(left=0, bottom=0, right=0.8, top=1)
    
    # simplest 'mappable'
    h = ax.pcolormesh([[0, 1]])
    
    # create colorbar
    cbax = f.add_axes([0, 0, 0.1, 0.1])

    cbar = plt.colorbar(h, orientation='vertical', cax=cbax)

    func = _resize_colorbar_vert(cbax, ax, **kwargs)
    f.canvas.mpl_connect('draw_event', func)

    f.canvas.draw()

    return f, cbar


def test_resize_colorbar_vert():

    # test pad=0, size=0.2
    f, cbar = _easy_cbar_vert(size=0.2, pad=0)
    
    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.8, 0, 0.2 * 0.8, 1.]

    np.testing.assert_allclose(res, exp)

    # -----------------------------------------------------------

    f.subplots_adjust(left=0, bottom=0.1, right=0.8, top=0.9)
    f.canvas.draw()

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.8, 0.1, 0.2 * 0.8, 0.8]

    np.testing.assert_allclose(res, exp)
    
    plt.close()
    
    # ===========================================================

    # test pad=0, aspect=5
    f, cbar = _easy_cbar_vert(aspect=5, pad=0)
    
    pos = cbar.ax.get_position()
    
    # don't test width if aspect is given, but also test aspect
    res = [pos.x0, pos.y0, pos.height]
    exp = [0.8, 0, 1.]
    np.testing.assert_allclose(res, exp)

    assert cbar.ax.get_aspect() == 5
    
    plt.close()

    # ===========================================================

    # test pad=0, aspect=default (=20)
    f, cbar = _easy_cbar_vert(pad=0)
    
    pos = cbar.ax.get_position()
    
    # don't test width if aspect is given, but also test aspect
    res = [pos.x0, pos.y0, pos.height]
    exp = [0.8, 0, 1.]
    np.testing.assert_allclose(res, exp)

    assert cbar.ax.get_aspect() == 20
    
    plt.close()

    # ===========================================================

    # pad=0.05, size=0.1

    f, cbar = _easy_cbar_vert(size=0.1, pad=0.05)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.8 + 0.8 * 0.05, 0, 0.8 * 0.1, 1.]

    np.testing.assert_allclose(res, exp)
    
    plt.close()

    # ===========================================================

    # shift='symmetric', shrink=0.1 
    # --> colorbar is 10 % smaller, and centered

    f, cbar = _easy_cbar_vert(size=0.2, pad=0., shrink=0.1)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.8, 0.05, 0.2 * 0.8, 0.9]

    np.testing.assert_allclose(res, exp)
    
    plt.close()

    # ===========================================================

    # shift=0., shrink=0.1
    # --> colorbar is 10 % smaller, and aligned with the bottom

    f, cbar = _easy_cbar_vert(size=0.2, pad=0., shrink=0.1, shift=0.)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.8, 0.0, 0.2 * 0.8, 0.9]

    np.testing.assert_allclose(res, exp)
    
    plt.close()

    # ===========================================================
    
    # shift=0.1, shrink=None
    # --> colorbar is 10 % smaller, and aligned with the top

    f, cbar = _easy_cbar_vert(size=0.2, pad=0., shrink=None, shift=0.1)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.8, 0.1, 0.2 * 0.8, 0.9]

    np.testing.assert_allclose(res, exp)
    
    plt.close()

# =============================================================================
# =============================================================================

def _easy_cbar_horz(**kwargs):

    f, ax = plt.subplots()
    
    f.subplots_adjust(left=0, bottom=0.2, right=1, top=1)

    # simplest 'mappable'
    h = ax.pcolormesh([[0, 1]])
    
    # create colorbar
    cbax = f.add_axes([0, 0, 0.1, 0.1])

    cbar = plt.colorbar(h, orientation='horizontal', cax=cbax)

    func = _resize_colorbar_horz(cbax, ax, **kwargs)
    f.canvas.mpl_connect('draw_event', func)

    f.canvas.draw()

    return f, cbar


def test_resize_colorbar_horz():


    # test pad=0, size=0.2
    f, cbar = _easy_cbar_horz(size=0.2, pad=0)
    
    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0, 0.2 * (1 - 0.8), 1, 0.2 * 0.8]

    # adding atol because else 5e-17 != 0
    np.testing.assert_allclose(res, exp, atol=1e-08)
    
    # -----------------------------------------------------------

    f.subplots_adjust(left=0.1, bottom=0.2, right=0.9, top=1)
    f.canvas.draw()

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.1, 0.2 * (1 - 0.8), 0.8, 0.2 * 0.8]

    np.testing.assert_allclose(res, exp, atol=1e-08)
    
    plt.close()
    
    # ===========================================================

    # test pad=0, aspect=5
    f, cbar = _easy_cbar_horz(aspect=5, pad=0)
    
    pos = cbar.ax.get_position()
    
    # don't test width if aspect is given, but also test aspect
    res = [pos.x0, pos.y0 + pos.height, pos.width]
    exp = [0., 0.2, 1.]
    np.testing.assert_allclose(res, exp)

    assert cbar.ax.get_aspect() == 1. / 5
    
    plt.close()

    # ===========================================================

    # test pad=0, aspect=default (=20)
    f, cbar = _easy_cbar_horz(pad=0)
    
    pos = cbar.ax.get_position()
    
    # don't test width if aspect is given, but also test aspect
    res = [pos.x0, pos.y0 + pos.height, pos.width]
    exp = [0., 0.2, 1.]
    np.testing.assert_allclose(res, exp)

    assert cbar.ax.get_aspect() == 1. / 20
    
    plt.close()

    # ===========================================================

    # pad=0.05, size=0.1

    f, cbar = _easy_cbar_horz(size=0.1, pad=0.05)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.0, 0.2 - 0.15 * 0.8, 1, 0.1 * 0.8]

    np.testing.assert_allclose(res, exp, atol=1e-08)
    
    plt.close()

    # ===========================================================

    # shift='symmetric', shrink=0.1 
    # --> colorbar is 10 % smaller, and centered

    f, cbar = _easy_cbar_horz(size=0.2, pad=0., shrink=0.1)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.05, 0.2 * (1 - 0.8), 0.9, 0.2 * 0.8]

    np.testing.assert_allclose(res, exp, atol=1e-08)
    
    plt.close()

    # ===========================================================

    # shift=0., shrink=0.1
    # --> colorbar is 10 % smaller, and aligned with lhs

    f, cbar = _easy_cbar_horz(size=0.2, pad=0., shrink=0.1, shift=0.)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.0, 0.2 * (1 - 0.8), 0.9, 0.2 * 0.8]

    np.testing.assert_allclose(res, exp, atol=1e-08)
    
    plt.close()

    # ===========================================================
    
    # shift=0.1, shrink=None
    # --> colorbar is 10 % smaller, and aligned with rhs

    f, cbar = _easy_cbar_horz(size=0.2, pad=0., shrink=None, shift=0.1)

    pos = cbar.ax.get_position()
    res = [pos.x0, pos.y0, pos.width, pos.height]
    exp = [0.1, 0.2 * (1 - 0.8), 0.9, 0.2 * 0.8]

    np.testing.assert_allclose(res, exp, atol=1e-08)
    
    plt.close()

# =============================================================================


def test_colorbar():

    # only test high level functionality
    f1, ax1 = plt.subplots()
    h = ax1.pcolormesh([[0, 1]])


    with raises(AssertionError):
        colorbar(h, ax1, orientation='wrong')

    with raises(RuntimeError):
        colorbar(h, ax1, anchor=5)

    with raises(RuntimeError):
        colorbar(h, ax1, panchor=5)

    with raises(AssertionError):
        f2, ax2 = plt.subplots()
        colorbar(h, ax1, ax2)


# =============================================================================

def test_get_cbax():

    f, ax = plt.subplots()

    cbax = _get_cbax(f)
    
    assert isinstance(cbax, plt.Axes)

    assert len(f.get_axes()) == 2

    _get_cbax(f)
    assert len(f.get_axes()) == 3

    _get_cbax(f)
    assert len(f.get_axes()) == 4

    _get_cbax(f)
    assert len(f.get_axes()) == 5


mplotutils/test/test_sample_data_3d.py
from mplotutils import sample_data_3d

import numpy as np

def test_data_shape():

    nlons = 10
    nlats = 20
    
    lon, lat, data = sample_data_3d(nlons, nlats)

    assert len(lon) == nlons
    assert len(lat) == nlats

    assert data.shape == (nlats, nlons)

    nlons = 5
    nlats = 10

    lon, lat, data = sample_data_3d(nlons, nlats)

    assert len(lon) == nlons
    assert len(lat) == nlats

    assert data.shape == (nlats, nlons)


def test_lat():

    __, lat, __ = sample_data_3d(36, 9)

    expected_lat = np.arange(-80, 81, 20)
    assert np.allclose(lat, expected_lat)


def test_lon():

    lon, __, __ = sample_data_3d(36, 9)

    expected_lon = np.arange(0, 351, 10)
    assert np.allclose(lon, expected_lon)

mplotutils/mpl_utils.py
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import six
from matplotlib.colors import from_levels_and_colors

# =============================================================================


def from_levels_and_cmap(levels, cmap, extend="neither"):
    """
    create mpl colormap from levels and the name of the colorbar

    Parameters
    ----------
    levels : sequence of numbers
        The quantization levels used to construct the :class:`BoundaryNorm`.
        Values ``v`` are quantizized to level ``i`` if
        ``lev[i] <= v < lev[i+1]``.
    cmap : string
        Valid colormap identifier.
    extend : {'neither', 'min', 'max', 'both'}, optional
        The behaviour when a value falls out of range of the given levels.
        See :func:`~matplotlib.pyplot.contourf` for details.

    ..note::
      Adapted from xarray.

    """
    if np.isscalar(levels):
        raise ValueError("'levels' must be a list of levels")

    if extend == "both":
        ext_n = 2
    elif extend in ["min", "max"]:
        ext_n = 1
    else:
        ext_n = 0

    # subtract 1 because there is one less level than numbers
    n_colors = len(levels) + ext_n - 1

    pal = _color_palette(cmap, n_colors)

    cmap, norm = from_levels_and_colors(levels, pal, extend=extend)

    return cmap, norm


# -----------------------------------------------------------------------------

# _color_palette is adapted from xarray:
# https://github.com/pydata/xarray/blob/v0.10.2/xarray/plot/utils.py#L110
# Used under the terms of xarrays's license, see licenses/XARRAY_LICENSE.


def _color_palette(cmap, n_colors):
    import matplotlib.pyplot as plt
    from matplotlib.colors import ListedColormap

    colors_i = np.linspace(0, 1.0, n_colors)
    if isinstance(cmap, (list, tuple)):
        # we have a list of colors
        cmap = ListedColormap(cmap, N=n_colors)
        pal = cmap(colors_i)
    elif isinstance(cmap, six.string_types):
        # we have some sort of named palette
        try:
            # is this a matplotlib cmap?
            cmap = plt.get_cmap(cmap)
            pal = cmap(colors_i)
        except ValueError:
            # ValueError happens when mpl doesn't like a colormap, try seaborn
            try:
                from seaborn.apionly import color_palette

                pal = color_palette(cmap, n_colors=n_colors)
            except (ValueError, ImportError):
                # or maybe we just got a single color as a string
                cmap = ListedColormap([cmap], N=n_colors)
                pal = cmap(colors_i)
    else:
        # cmap better be a LinearSegmentedColormap (e.g. viridis)
        pal = cmap(colors_i)

    return pal


# =============================================================================


def set_map_layout(axes, width=17.0, nrow=None, ncol=None):
    """
    set figure height, given width

    Needs to be called after all plotting is done.

    Parameters
    ----------
    axes : ndarray of (Geo)Axes
        Array with all axes of the figure.
    width : float, default: 17
        Width of the full figure in cm.
    nrow : integer, default: None
        manually set the number of rows of subplots. Good when using gridspec.
        However, subplots must span the same number of gridspec rows & columns.
        Either none or both of 'nrow' and 'ncol' must be set.
    ncol : integer, default: None
        As nrow but for the number of rows.

    ..note: only works if all the axes have the same aspect ratio.
    """

    if (nrow is None and ncol is not None) or (nrow is not None and ncol is None):
        raise ValueError("Must set none or both of 'nrow' and 'ncol'")

    if isinstance(axes, plt.Axes):
        ax = axes
    else:
        # assumes the first of the axes is representative for all
        ax = axes.flat[0]

    # read figure data
    f = ax.get_figure()

    bottom = f.subplotpars.bottom
    top = f.subplotpars.top
    left = f.subplotpars.left
    right = f.subplotpars.right
    hspace = f.subplotpars.hspace
    wspace = f.subplotpars.wspace

    # data ratio is the aspect
    aspect = ax.get_data_ratio()

    if nrow is None and ncol is None:
        # get geometry tells how many subplots there are
        nrow, ncol, __ = ax.get_geometry()

    # width of one plot, taking into account
    # left * wf, (1-right) * wf, ncol * wp, (1-ncol) * wp * wspace
    wp = (width - width * (left + (1 - right))) / (ncol + (ncol - 1) * wspace)

    # height of one plot
    hp = wp * aspect

    # height of figure
    height = (hp * (nrow + ((nrow - 1) * hspace))) / (1.0 - (bottom + (1 - top)))

    f.set_figwidth(width / 2.54)
    f.set_figheight(height / 2.54)


# =============================================================================


def _get_label_attr(labelpad, size, weight):

    if labelpad is None:
        labelpad = mpl.rcParams["axes.labelpad"]

    if size is None:
        size = mpl.rcParams["axes.labelsize"]

    if weight is None:
        weight = mpl.rcParams["axes.labelweight"]

    return labelpad, size, weight


mplotutils/colorbar_utils.py
import matplotlib.pyplot as plt


def colorbar(
    mappable,
    ax1,
    ax2=None,
    orientation="vertical",
    aspect=None,
    size=None,
    pad=None,
    shift="symmetric",
    shrink=None,
    **kwargs
):
    """
    automatically resize colorbars on draw

    See below for Example

    Parameters
    ----------
    mappable : handle
        The return value of 'ax.contourf', 'ax.pcolormesh' etc to
        which the colorbar applies.
    ax1 : Axes
        Axes to adjust the colorbar to.
    ax2 : Axes, optional
        If the colorbar should span more than one Axes. Default: None.
    orientation : 'vertical' | 'horizontal', optional
        Orientation of the colorbar. Default: 'vertical'.
    aspect : float, optional
        The ratio of long to short dimensions of the colorbar. Default 20.
        Mutually exclusive with 'size'.
    size : float
        Width of the colorbar as fraction of the axes width (vertical) or
        height (horizontal). Mutually exclusive with 'aspect'. Default: None.
    pad : float
        Distance of the colorbar to the axes in Figure coordinates.
         Default: 0.05 (vertical) or 0.15 (horizontal).
    shift : 'symmetric' or float in 0..1
        Fraction of the total height that the colorbar is shifted upwards.
        See Note. Default: 'symmetric'
    shrink : None or float in 0..1
        Fraction of the total height that the colorbar is shrunk.
        See Note. Default: None.
    **kwargs : keyword arguments
        colorbar properties
        ============  ====================================================
        Property      Description
        ============  ====================================================
        *extend*      [ 'neither' | 'both' | 'min' | 'max' ]
                    If not 'neither', make pointed end(s) for out-of-
                    range values.  These are set for a given colormap
                    using the colormap set_under and set_over methods.
        *extendfrac*  [ *None* | 'auto' | length | lengths ]
                    If set to *None*, both the minimum and maximum
                    triangular colorbar extensions with have a length of
                    5% of the interior colorbar length (this is the
                    default setting). If set to 'auto', makes the
                    triangular colorbar extensions the same lengths as
                    the interior boxes (when *spacing* is set to
                    'uniform') or the same lengths as the respective
                    adjacent interior boxes (when *spacing* is set to
                    'proportional'). If a scalar, indicates the length
                    of both the minimum and maximum triangular colorbar
                    extensions as a fraction of the interior colorbar
                    length. A two-element sequence of fractions may also
                    be given, indicating the lengths of the minimum and
                    maximum colorbar extensions respectively as a
                    fraction of the interior colorbar length.
        *extendrect*  [ *False* | *True* ]
                    If *False* the minimum and maximum colorbar extensions
                    will be triangular (the default). If *True* the
                    extensions will be rectangular.
        *spacing*     [ 'uniform' | 'proportional' ]
                    Uniform spacing gives each discrete color the same
                    space; proportional makes the space proportional to
                    the data interval.
        *ticks*       [ None | list of ticks | Locator object ]
                    If None, ticks are determined automatically from the
                    input.
        *format*      [ None | format string | Formatter object ]
                    If None, the
                    :class:`~matplotlib.ticker.ScalarFormatter` is used.
                    If a format string is given, e.g., '%.3f', that is
                    used. An alternative
                    :class:`~matplotlib.ticker.Formatter` object may be
                    given instead.
        *drawedges*   [ False | True ] If true, draw lines at color
                    boundaries.
        ============  ====================================================

    Examples
    --------
    import matplotlib.pyplot as plt
    import mplotutils as mpu
    import cartopy.crs as ccrs

    # example with 1 axes

    f = plt.figure()
    ax = plt.axes(projection=ccrs.PlateCarree())
    h = ax.pcolormesh([[0, 1]])

    ax.coastlines()

    mpu.colorbar(h, ax)

    ax.set_global()

    # =========================
    # example with 2 axes

    f, axes = plt.subplots(2, 1, subplot_kw=dict(projection=ccrs.Robinson()))

    for ax in axes:
        ax.coastlines()
        ax.set_global()
        h = ax.pcolormesh([[0, 1]])

    cbar = mpu.colorbar(h, axes[0], axes[1])

    cbar.set_label('[°C]', labelpad=10)

    # =========================
    # example with 3 axes & 2 colorbars

    f, axes = plt.subplots(3, 1, subplot_kw=dict(projection=ccrs.Robinson()))

    for ax in axes:
        ax.coastlines()
        ax.set_global()

    h0 = ax.pcolormesh([[0, 1]])
    h1 = ax.pcolormesh([[0, 1]])
    h2 = ax.pcolormesh([[0, 1]], cmap='Blues')

    cbar = mpu.colorbar(h, axes[0], axes[1], size=0.05)

    cbar = mpu.colorbar(h, axes[2], size=0.05)

    plt.draw()

    Notes
    -----
    shift='symmetric', shrink=None  -> colorbar extends over the whole height
    shift='symmetric', shrink=0.1   -> colorbar is 10 % smaller, and centered
    shift=0., shrink=0.1            -> colorbar is 10 % smaller, and aligned
                                       with the bottom
    shift=0.1, shrink=None          -> colorbar is 10 % smaller, and aligned
                                       with the top

    See Also
    --------
    _resize_colorbar_horz
    """

    orientations = ("vertical", "horizontal")
    if orientation not in orientations:
        raise ValueError("orientation must be 'vertical' or 'horizontal'")

    k = kwargs.keys()
    if ("anchor" in k) or ("panchor" in k):
        msg = (
            "'anchor' and 'panchor' keywords not " "supported, use 'shrink' and 'shift'"
        )
        raise ValueError(msg)

    # ensure 'ax' does not end up in plt.colorbar(**kwargs)
    if "ax" in k:
        # assume it is ax2 (it can't be ax1)
        ax2 = kwargs.pop("ax")

    f = ax1.get_figure()

    if ax2 is not None:
        f2 = ax2.get_figure()
        assert f == f2, "'ax1' and 'ax2' must belong to the same figure"

    cbax = _get_cbax(f)

    cbar = plt.colorbar(mappable, orientation=orientation, cax=cbax, **kwargs)

    if orientation == "vertical":
        func = _resize_colorbar_vert(
            cbax,
            ax1,
            ax2=ax2,
            aspect=aspect,
            size=size,
            pad=pad,
            shift=shift,
            shrink=shrink,
        )
    else:
        func = _resize_colorbar_horz(
            cbax,
            ax1,
            ax2=ax2,
            aspect=aspect,
            size=size,
            pad=pad,
            shift=shift,
            shrink=shrink,
        )

    f.canvas.mpl_connect("draw_event", func)
    f.canvas.draw()

    return cbar


# ========================================================================


def _get_cbax(f):

    # when using f.add_axes(rect) with the same rect twice
    # it is the same axes, so we have to change rect
    # slightly each time this func is called

    n_axes = len(f.get_axes())
    pos_incr = n_axes / 10.0
    return f.add_axes([0, 0, 0.1, 0.1 + pos_incr])


def _resize_colorbar_vert(
    cbax,
    ax1,
    ax2=None,
    aspect=None,
    size=None,
    pad=None,
    shift="symmetric",
    shrink=None,
):
    """
    automatically resize colorbars on draw

    see 'colorbar'

    Examples
    --------
    import matplotlib.pyplot as plt
    import mplotutils as mpu
    import cartopy.crs as ccrs

    f = plt.figure()
    ax = plt.axes(projection=ccrs.PlateCarree())
    h = ax.pcolormesh([[0, 1]])

    ax.coastlines()

    cbax = f.add_axes([0, 0, 0.1, 0.1])
    cbar = plt.colorbar(h, orientation='vertical', cax=cbax)

    func = mpu.colorbar_utils_resize_colorbar_vert(cbax, ax)
    f.canvas.mpl_connect('draw_event', func)

    ax.set_global()

    plt.draw()

    See Also
    --------
    _resize_colorbar_horz
    """

    shift, shrink = _parse_shift_shrink(shift, shrink)

    size, aspect, pad = _parse_size_aspect_pad(size, aspect, pad, "vertical")

    # swap axes if ax1 is above ax2
    if ax2 is not None:
        posn1 = ax1.get_position()
        posn2 = ax2.get_position()

        ax1, ax2 = (ax1, ax2) if posn1.y0 < posn2.y0 else (ax2, ax1)

    if aspect is not None:
        anchor = (0, 0.5)
        cbax.set_aspect(aspect, anchor=anchor, adjustable="box")

    # inner function is called by event handler
    def inner(event=None):

        posn1 = ax1.get_position()

        # determine total height of all axes
        if ax2 is not None:
            posn2 = ax2.get_position()
            full_height = posn2.y0 - posn1.y0 + posn2.height
        else:
            full_height = posn1.height

        pad_scaled = pad * posn1.width
        size_scaled = size * posn1.width

        # calculate position
        left = posn1.x0 + posn1.width + pad_scaled
        bottom = posn1.y0 + shift * full_height
        height = full_height - shrink * full_height
        # ignored if aspect is set
        width = size_scaled

        pos = [left, bottom, width, height]

        cbax.set_position(pos)

    return inner


# ====================================


def _resize_colorbar_horz(
    cbax,
    ax1,
    ax2=None,
    aspect=None,
    size=None,
    pad=None,
    shift="symmetric",
    shrink=None,
):
    """
    automatically resize colorbars on draw

    see 'colorbar'

    Examples
    --------
    import matplotlib.pyplot as plt
    import mplotutils as mpu
    import cartopy.crs as ccrs

    f = plt.figure()
    ax = plt.axes(projection=ccrs.PlateCarree())

    ax.coastlines()

    cbax = f.add_axes([0, 0, 0.1, 0.1])
    cbar = plt.colorbar(h, orientation='horizontal', cax=cbax)

    func = mpu.colorbar_utils._resize_colorbar_horz(cbax, ax)
    f.canvas.mpl_connect('draw_event', func)

    ax.set_global()

    plt.draw()

    See Also
    --------
    _resize_colorbar_vert
    """

    shift, shrink = _parse_shift_shrink(shift, shrink)

    size, aspect, pad = _parse_size_aspect_pad(size, aspect, pad, "horizontal")

    if ax2 is not None:
        posn1 = ax1.get_position()
        posn2 = ax2.get_position()

        # swap axes if ax1 is right of ax2
        ax1, ax2 = (ax1, ax2) if posn1.x0 < posn2.x0 else (ax2, ax1)

    if aspect is not None:
        aspect = 1 / aspect
        anchor = (0.5, 1.0)
        cbax.set_aspect(aspect, anchor=anchor, adjustable="box")

    def inner(event=None):

        posn1 = ax1.get_position()

        if ax2 is not None:
            posn2 = ax2.get_position()
            full_width = posn2.x0 - posn1.x0 + posn2.width
        else:
            full_width = posn1.width

        pad_scaled = pad * posn1.height
        size_scaled = size * posn1.height

        left = posn1.x0 + shift * full_width
        bottom = posn1.y0 - (pad_scaled + size_scaled)
        width = full_width - shrink * full_width

        # ignored if aspect is set
        height = size_scaled

        pos = [left, bottom, width, height]

        cbax.set_position(pos)

    return inner


# ====================================


def _parse_shift_shrink(shift, shrink):

    if shift == "symmetric":
        if shrink is None:
            shrink = 0

        shift = shrink / 2.0

    else:
        if shrink is None:
            shrink = shift

    assert (shift >= 0.0) & (shift <= 1.0), "'shift' must be in 0...1"
    assert (shrink >= 0.0) & (shrink <= 1.0), "'shrink' must be in 0...1"

    if shift > shrink:
        msg = (
            "Warning: 'shift' is larger than 'shrink', colorbar\n"
            "will extend beyond the axes!"
        )
        print(msg)

    return shift, shrink


# ==================================================================================================


def _parse_size_aspect_pad(size, aspect, pad, orientation):

    if (size is not None) and (aspect is not None):
        raise ValueError("you can only pass one of 'aspect' and 'size'")

    # default is aspect=20
    if (size is None) and (aspect is None):
        aspect = 20

    # we need a large size so it is not limiting for set_aspect
    if aspect is not None:
        size = 10

    # default mpl setting
    if pad is None:
        pad = 0.05 if orientation == "vertical" else 0.15

    return size, aspect, pad
